File: src/main/java/io/github/intellij/dlanguage/stubs/types/ClassDeclarationStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.named.DlangClassDeclaration;
import io.github.intellij.dlanguage.psi.impl.named.DlangClassDeclarationImpl;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.DlangClassDeclarationStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

public class ClassDeclarationStubElementType extends DNamedStubElementType<DlangClassDeclarationStub, DlangClassDeclaration> {
    public ClassDeclarationStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DlangClassDeclaration createPsi(@NotNull final DlangClassDeclarationStub stub) {
        return new DlangClassDeclarationImpl(stub, this);
    }

    @NotNull
    @Override
    public DlangClassDeclarationStub createStub(@NotNull final DlangClassDeclaration psi, final StubElement parentStub) {
        return new DlangClassDeclarationStub(parentStub, this, psi.getName(), psi.getAttributes());
    }

    @Override
    public void serialize(@NotNull final DlangClassDeclarationStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }

    @NotNull
    @Override
    public DlangClassDeclarationStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DlangClassDeclarationStub(parentStub, this, dataStream.readName(),
            DAttributes.Companion.read(dataStream));
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/ClassDeclarationStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DlangClassDeclarationImpl
import io.github.intellij.dlanguage.psi.named.DLanguageClassDeclaration
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageClassDeclarationStub
import java.io.IOException

class ClassDeclarationStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageClassDeclarationStub, DLanguageClassDeclaration>(debugName) {
    override fun createPsi(stub: DLanguageClassDeclarationStub): DLanguageClassDeclaration {
        return DlangClassDeclarationImpl(stub, this)
    }

    override fun createStub(psi: DLanguageClassDeclaration, parentStub: StubElement<*>?): DLanguageClassDeclarationStub {
        return DLanguageClassDeclarationStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageClassDeclarationStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageClassDeclarationStub {
        return DLanguageClassDeclarationStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
